### ISSUE TITLE
Very long reactor stalls after booting a node with 1000 keyspaces

### ISSUE THREAD

**Issue by roydahan** [scylladb/scylla#4133]
*Installation details*
Scylla version (or git commit hash): 3.0.1-0.20190120.3c4f8cf6e
Cluster size: 6 x i3.8xlarge
OS (RHEL/CentOS/Ubuntu/AWS AMI): AWS AMI (ami-077bbd05e075e0a62)

During multi-keyspaces test, when DestoryThenRebuild nemesis (stopping node, deleting few files, starting the node and running 'nodetool rebuild') tried to run on a node with 1000 keyspaces, the node had very long reactor stalls after booting.

Stalls started from 1000 ms and reached to **128000 ms**.
At some point the client and other nodes had connections timeouts to this node:
`rpc - client 172.16.99.52: fail to connect: Connection timed out`

` [shard 6] storage_proxy - Exception when communicating with 172.16.99.52: seastar::semaphore_timed_out (Semaphore timedout)`

Backtrace:
```
Reactor stalled for 128000 ms on shard 0.
 Backtrace:
 0x00000000006a8542
 0x00000000005bfc6c
 0x00000000005c0126
 /lib64/libpthread.so.0+0x000000000000f5cf
 0x0000000002fc57cf
 0x0000000002fcb924
 0x00000000035dfa0e
 0x0000000003644c7f
 0x0000000003646e8e
 0x0000000003d5ea3b
 0x0000000000af7db3
 0x0000000000b0cc7a
 0x0000000000b2c99d
 0x0000000000b2f590
 0x0000000000b30ece
 0x00000000005a101e
 0x00000000005a11e1
 0x000000000066f877
 0x00000000007ad583
 0x0000000000534767
 /lib64/libc.so.6+0x00000000000223d4
 0x000000000059af1d
```

Decode:
```
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/src/debug/scylla-3.0.1/seastar/util/backtrace.hh:56
seastar::backtrace_buffer::append_backtrace() at /usr/src/debug/scylla-3.0.1/seastar/core/reactor.cc:410
 (inlined by) print_with_backtrace at /usr/src/debug/scylla-3.0.1/seastar/core/reactor.cc:431
seastar::reactor::block_notifier(int) at /usr/src/debug/scylla-3.0.1/seastar/core/reactor.cc:749
?? ??:0
wrapping_range<dht::token>::wrapping_range(wrapping_range<dht::token> const&) at /usr/src/debug/scylla-3.0.1/./range.hh:61
 (inlined by) nonwrapping_range<dht::token>::operator wrapping_range<dht::token>() const & at /usr/src/debug/scylla-3.0.1/./range.hh:458
 (inlined by) dht::range_streamer::get_all_ranges_with_sources_for(seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::vector<nonwrapping_range<dht::token>, std::allocator<nonwrapping_range<dht::token> > >) at /usr/src/debug/scylla-3.0.1/dht/range_streamer.cc:118
dht::range_streamer::add_ranges(seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::vector<nonwrapping_range<dht::token>, std::allocator<nonwrapping_range<dht::token> > >) at /usr/src/debug/scylla-3.0.1/dht/range_streamer.cc:236
operator() at /usr/src/debug/scylla-3.0.1/service/storage_service.cc:2470
operator() at /usr/src/debug/scylla-3.0.1/service/storage_service.hh:2219
 (inlined by) operator() at /usr/src/debug/scylla-3.0.1/seastar/core/sharded.hh:332
 (inlined by) apply<seastar::sharded<T>::invoke_on(unsigned int, Func&&) [with Func = service::storage_service::run_with_api_lock(seastar::sstring, Func&&) [with Func = service::storage_service::rebuild(seastar::sstring)::<lambda(service::storage_service&)>]::<lambda(service::storage_service&)>; Ret = seastar::future<>; Service = service::storage_service]::<lambda()>&> at /usr/src/debug/scylla-3.0.1/seastar/core/future.hh:1399
submit_to<seastar::sharded<T>::invoke_on(unsigned int, Func&&) [with Func = service::storage_service::run_with_api_lock(seastar::sstring, Func&&) [with Func = service::storage_service::rebuild(seastar::sstring)::<lambda(service::storage_service&)>]::<lambda(service::storage_service&)>; Ret = seastar::future<>; Service = service::storage_service]::<lambda()> > at /usr/src/debug/scylla-3.0.1/seastar/core/reactor.hh:1247
 (inlined by) invoke_on<service::storage_service::run_with_api_lock(seastar::sstring, Func&&) [with Func = service::storage_service::rebuild(seastar::sstring)::<lambda(service::storage_service&)>]::<lambda(service::storage_service&)> > at /usr/src/debug/scylla-3.0.1/seastar/core/sharded.hh:333
 (inlined by) run_with_api_lock<service::storage_service::rebuild(seastar::sstring)::<lambda(service::storage_service&)> > at /usr/src/debug/scylla-3.0.1/service/storage_service.hh:2220
 (inlined by) service::storage_service::rebuild(seastar::basic_sstring<char, unsigned int, 15u, true>) at /usr/src/debug/scylla-3.0.1/service/storage_service.cc:2479
operator() at /usr/src/debug/scylla-3.0.1/api/storage_service.cc:635
 (inlined by) _M_invoke at /opt/scylladb/include/c++/7/bits/std_function.h:302
std::function<seastar::future<seastar::json::json_return_type> (std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >)>::operator()(std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >) const at /opt/scylladb/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::httpd::function_handler::function_handler(std::function<seastar::future<seastar::json::json_return_type> (std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >)> const&)::{lambda(std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >)#1}::operator()(std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >) const at /usr/src/debug/scylla-3.0.1/seastar/http/function_handlers.hh:102
 (inlined by) std::_Function_handler<seastar::future<std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> > > (std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >), seastar::httpd::function_handler::function_handler(std::function<seastar::future<seastar::json::json_return_type> (std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >)> const&)::{lambda(std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >)#1}>::_M_invoke(std::_Any_data const&, std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >&&, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >&&) at /opt/scylladb/include/c++/7/bits/std_function.h:302
std::function<seastar::future<std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> > > (std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >)>::operator()(std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >) const at /opt/scylladb/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::httpd::function_handler::handle(seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >) at /usr/src/debug/scylla-3.0.1/seastar/http/function_handlers.hh:118
 (inlined by) seastar::httpd::routes::handle(seastar::basic_sstring<char, unsigned int, 15u, true> const&, std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >, std::unique_ptr<seastar::httpd::reply, std::default_delete<seastar::httpd::reply> >) at /usr/src/debug/scylla-3.0.1/seastar/http/routes.cc:94
seastar::httpd::connection::generate_reply(std::unique_ptr<seastar::httpd::request, std::default_delete<seastar::httpd::request> >) at /usr/src/debug/scylla-3.0.1/seastar/http/httpd.cc:231
operator() at /usr/src/debug/scylla-3.0.1/seastar/http/httpd.cc:173
 (inlined by) apply at /usr/src/debug/scylla-3.0.1/seastar/core/apply.hh:35
 (inlined by) apply<seastar::httpd::connection::read_one()::<lambda()> mutable::<lambda()> > at /usr/src/debug/scylla-3.0.1/seastar/core/apply.hh:43
 (inlined by) apply<seastar::httpd::connection::read_one()::<lambda()> mutable::<lambda()> > at /usr/src/debug/scylla-3.0.1/seastar/core/future.hh:1389
 (inlined by) then<seastar::httpd::connection::read_one()::<lambda()> mutable::<lambda()> > at /usr/src/debug/scylla-3.0.1/seastar/core/future.hh:952
 (inlined by) operator() at /usr/src/debug/scylla-3.0.1/seastar/http/httpd.cc:171
apply at /usr/src/debug/scylla-3.0.1/seastar/core/apply.hh:35
 (inlined by) apply<seastar::httpd::connection::read_one()::<lambda()> > at /usr/src/debug/scylla-3.0.1/seastar/core/apply.hh:43
 (inlined by) apply<seastar::httpd::connection::read_one()::<lambda()> > at /usr/src/debug/scylla-3.0.1/seastar/core/future.hh:1389
 (inlined by) operator()<seastar::future_state<> > at /usr/src/debug/scylla-3.0.1/seastar/core/future.hh:963
 (inlined by) run_and_dispose at /usr/src/debug/scylla-3.0.1/seastar/core/future.hh:414
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/src/debug/scylla-3.0.1/seastar/core/reactor.cc:2695
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-3.0.1/seastar/core/reactor.cc:3118
seastar::reactor::run_some_tasks() at /usr/src/debug/scylla-3.0.1/seastar/util/log.hh:313
 (inlined by) seastar::reactor::run() at /usr/src/debug/scylla-3.0.1/seastar/core/reactor.cc:3265
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at /usr/src/debug/scylla-3.0.1/seastar/core/app-template.cc:185
main at /usr/src/debug/scylla-3.0.1/main.cc:859
```

**Comment by avikivity** [comment]
! /cc @asias !

**Comment by asias** [comment]
Nice stall.  

We already yield in the for loop. How come it stalled 128s?

101 std::unordered_map<dht::token_range, std::vector<inet_address>>
102 range_streamer::get_all_ranges_with_sources_for(const sstring& keyspace_name, dht::token_range_vector desired_ranges) {
103     logger.debug("{} ks={}", __func__, keyspace_name);
104 
105     auto& ks = _db.local().find_keyspace(keyspace_name);
106     auto& strat = ks.get_replication_strategy();
107 
108     auto tm = _metadata.clone_only_token_map();
109     auto range_addresses = strat.get_range_addresses(tm);
110 
111     logger.debug("keyspace={}, desired_ranges.size={}, range_addresses.size={}", keyspace_name, desired_ranges.size(), range_addresses.size());
112 
113     std::unordered_map<dht::token_range, std::vector<inet_address>> range_sources;
114     auto& snitch = locator::i_endpoint_snitch::get_local_snitch_ptr();
115     for (auto& desired_range : desired_ranges) {
116         auto found = false;
117         for (auto& x : range_addresses) {
118             if (need_preempt()) {
119                 seastar::thread::yield();
120             }
121             const range<token>& src_range = x.first;
122             if (src_range.contains(desired_range, dht::tri_compare)) {
123                 std::vector<inet_address>& addresses = x.second;
124                 auto preferred = snitch->get_sorted_list_by_proximity(_address, addresses);
125                 for (inet_address& p : preferred) {
126                     range_sources[desired_range].push_back(p);
127                 }
128                 found = true;
129             }
130         }
131 
132         if (!found) {
133             throw std::runtime_error(format("No sources found for {}", desired_range));
134         }
135     }
136 
137     return range_sources;
138 }

**Comment by asias** [comment]
I created 500 keyspaces and ran nodetool rebuild. I did not see any stalls locally.

```
for i in `seq 0 500`;do ks="ks$i" ;echo "Create $ks" ;cassandra-stress write no-warmup n=1 cl=ONE -schema 'replication(factor=2)' keyspace=$ks -pop seq=1..10 -rate threads=1 -node 127.0.0.1;done
```

**Comment by roydahan** [comment]
@asias, I can give you access to this cluster if you want.

**Comment by roydahan** [comment]
@asias, do you need the host or not?
I'll need to purge them soon.

**Comment by asias** [comment]
> @asias, do you need the host or not?
> I'll need to purge them soon.

I tried to reproduce myself first and failed. So I still need your cluster. 

You wrote: 
```
During multi-keyspaces test, when DestoryThenRebuild nemesis (stopping node, deleting few files, starting the node and running 'nodetool rebuild') tried to run on a node with 1000 keyspaces, the node had very long reactor stalls after booting.
```

Could you tell me exactly when did the stall happen, during  the node restart or during the nodetool rebuild?

**Comment by roydahan** [comment]
It happened right after the node rebuild, but I thought it's related to the reboot.
I'm sending you the server details in email, you can issue a nodetool rebuild without reboot and check if it reproduces.

**Comment by asias** [comment]
On which node the rebuild and stall happened?

**Comment by asias** [comment]
OK. I found it. It is supposed to be node 172-16-99-52.

**Comment by asias** [comment]
I managed to reproduce with nodetool rebuild:
```
[centos@ip-172-16-99-52 ~]$ journalctl -u scylla-server --since '09:30:29'|grep -i stall
Jan 28 09:30:50 ip-172-16-99-52.ec2.internal scylla[14909]: Reactor stalled for 1000 ms on shard 0.
Jan 28 09:30:51 ip-172-16-99-52.ec2.internal scylla[14909]: Reactor stalled for 2000 ms on shard 0.
Jan 28 09:30:53 ip-172-16-99-52.ec2.internal scylla[14909]: Reactor stalled for 4000 ms on shard 0.
Jan 28 09:30:57 ip-172-16-99-52.ec2.internal scylla[14909]: Reactor stalled for 8000 ms on shard 0.
Jan 28 09:31:05 ip-172-16-99-52.ec2.internal scylla[14909]: Reactor stalled for 16000 ms on shard 0.
Jan 28 09:31:53 ip-172-16-99-52.ec2.internal scylla[14909]: Reactor stalled for 64000 ms on shard 0.
Jan 28 09:32:57 ip-172-16-99-52.ec2.internal scylla[14909]: Reactor stalled for 128000 ms on shard 0.
[centos@ip-172-16-99-52 ~]$ journalctl -u scylla-server --since '09:30:29'|grep desired_ranges
Jan 28 09:30:49 ip-172-16-99-52.ec2.internal scylla[14909]:  [shard 0] range_streamer - keyspace=keyspace963, desired_ranges.size=758, range_addresses.size=1537
Jan 28 09:30:50 ip-172-16-99-52.ec2.internal scylla[14909]:  [shard 0] range_streamer - keyspace=keyspace81, desired_ranges.size=758, range_addresses.size=1537
Jan 28 09:31:19 ip-172-16-99-52.ec2.internal scylla[14909]:  [shard 0] range_streamer - keyspace=keyspace134, desired_ranges.size=758, range_addresses.size=1537
Jan 28 09:31:49 ip-172-16-99-52.ec2.internal scylla[14909]:  [shard 0] range_streamer - keyspace=keyspace191, desired_ranges.size=758, range_addresses.size=1537
Jan 28 09:32:19 ip-172-16-99-52.ec2.internal scylla[14909]:  [shard 0] range_streamer - keyspace=keyspace112, desired_ranges.size=758, range_addresses.size=1537
Jan 28 09:32:50 ip-172-16-99-52.ec2.internal scylla[14909]:  [shard 0] range_streamer - keyspace=keyspace456, desired_ranges.size=758, range_addresses.size=1537
Jan 28 09:33:20 ip-172-16-99-52.ec2.internal scylla[14909]:  [shard 0] range_streamer - keyspace=keyspace85, desired_ranges.size=758, range_addresses.size=1537
```

@avikivity The loop looks like below. Is it possible need_preempt() makes yield() not get called for some reason. 

```
  0     for (auto& desired_range : desired_ranges) {
  1         auto found = false;
  2         for (auto& x : range_addresses) {
  3             if (need_preempt()) {
  4                 seastar::thread::yield();
  5             }
  6             const range<token>& src_range = x.first;
  7             if (src_range.contains(desired_range, dht::tri_compare)) {
  8                 std::vector<inet_address>& addresses = x.second;
  9                 auto preferred = snitch->get_sorted_list_by_proximity(_address, addresses);
 10                 for (inet_address& p : preferred) {
 11                     range_sources[desired_range].push_back(p);
 12                 }
 13                 found = true;
 14             }
 15         }

```

**Comment by asias** [comment]
OK. Problem found. It turns out we did not backport the following fix to 3.0!

```
commit 8edf3defdf4bc0ece99ae11e60b6691e4cf12168
Author: Asias He <asias@scylladb.com>
Date:   Wed Sep 5 16:21:12 2018 +0800

    range_streamer: Futurize add_ranges

    It might take long time for get_all_ranges_with_sources_for and
    get_all_ranges_with_strict_sources_for to calculate which cause reactor
    stall. To fix, run them in a thread and yield. Those functions are used in
    the slow path, it is ok to yield more than needed.

    Fixes #3639
    
    Message-Id: <63aa7794906ac020c9d9b2984e1351a8298a249b.1536135617.git.asias@scylladb.com>
```

**Comment by asias** [comment]
@roydahan I do not need the cluster any more.

**Comment by roydahan** [comment]
Thx, killing it.

**Comment by glommer** [comment]
Is tha plan to backport it now ?

**Comment by asias** [comment]
I asked to backport this patch to 3.0 on Dec 26, 2018. Avi rejected and I asked again yesterday. I am not sure if his mind changes. 
@avikivity 

```
On Wed, Dec 26, 2018 at 3:41 PM Avi Kivity <avi@scylladb.com> wrote:

    On 12/26/18 3:43 AM, Asias He wrote:
>
>
>     On Thu, Dec 20, 2018 at 9:56 PM Commit Bot <bot@cloudius-systems.com> wrote:
>
>         From: Asias He <asias@scylladb.com>
>         Committer: Avi Kivity <avi@scylladb.com>
>         Branch: master
>
>
>     Backport to 3.0 ?
>

I'd rather not risk 3.0 any more. This doesn't fix a regression, and the call only happens O(1) times per repair, so it doesn't impact the 99th percentile.
```

**Comment by slivne** [comment]
the fix was backport to 3.0 45bb1ba1b70ef55b8094af98adcc61ec7741dc5b its part of 3.0.2

**Comment by slivne** [comment]
my bad was not backported wrong commit

**Comment by asias** [comment]
It is backported to 3.0 branch with commit 22f41f04baf870a85ca2b752ae6b92fd0901956e. Can we close it now?

**Comment by roydahan** [comment]
need to re-run the test in that case.

**Comment by slivne** [comment]
@roydahan ping

**Comment by roydahan** [comment]
Looks like it's fixed.
longevity with 1000 ks is runnning and already passed nemesis like "multiple hard reboots" without these stalls:

```
[54.224.191.17] [stdout] Mar 31 11:03:07 info   | Reactor stalled for 640 ms on shard 0.
[54.81.193.83] [stdout] Mar 31 11:04:31 info   | Reactor stalled for 718 ms on shard 0.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 26.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 3.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 10.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 22.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 19.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 20.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 12.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 15.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 24.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 4.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 5.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 21.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 27.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 29.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 2.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 8.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 28.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 14.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 23.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 6.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 17.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 16.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 7.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 25.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 9.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 1.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 18.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 11.
[34.238.159.150] [stdout] Mar 31 23:43:05 info   | Reactor stalled for 500 ms on shard 13.
```

**Comment by avikivity** [comment]
Fixed by 22f41f04baf870a85ca2b752ae6b92fd0901956e
